### PR TITLE
feat: [Settings Menu] Graphics and Sound sections

### LINKED ISSUE #579: Change game resolution from settings
### Problem
Current build does not have a way to change the game resolution to adapt it to your device specs, using limitless resolution of your display. This can hugely affect performance for devices with 2K or 4K resolutions or ultrawide screens.

Also the UI is not prepared to have ultra-ultrawide resolutions, and does not adapt properly in scaling.
image

### Solution
Show a dropdown of available resolutions in the graphics card, allowing only the ones for 16:9 and 16:10. Default resolution should be 1080 16:9
Also add a dropdown to control FPS cap on 30, 60, 120, 144 or limitless.
Add another dropdown to select between fullscreen and windowed.

### PR DESCRIPTION
**Pull request by sandrade-dcl**
## What does this PR change?
Fix #579 

**GRAPHICS**
- [x] Resolution.
- [x] FPS Limit.
- [x] Window Mode.

image

**SOUND**
- [x]  Master
- [x]  World SFX
- [x]  Avatar SFX
- [x]  UI SFX
- [x]  Music
- [x]  Chat Sounds

image

## How to test the changes?
1. Launch the explorer
2. Open the Settings Menu

## Our Code Review Standards
https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: Explorer/Assets/DCL/Settings/ModuleControllers/EnvironmentDistanceSettingsController.cs [hunk 1]
```diff
@@ -20,17 +21,23 @@ public EnvironmentDistanceSettingsController(SettingsSliderModuleView view, Land
 
             view.SliderView.Slider.onValueChanged.AddListener(SetEnvironmentDistanceSettings);
             SetEnvironmentDistanceSettings(view.SliderView.Slider.value);
+
+            landscapeData.OnDetailDistanceChanged += OnEnvironmentDistanceSettingsChangedFromOutside;
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/EnvironmentDistanceSettingsController.cs:25]: Can landscapeData change when we switch to and from worlds?

What is the lifecycle of EnvironmentDistanceSettingsController?

File: Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs [hunk 2]
```diff
@@ -10,33 +14,96 @@ public class GraphicsQualitySettingsController : SettingsFeatureController
...
+        public GraphicsQualitySettingsController(SettingsDropdownModuleView view, RealmPartitionSettingsAsset realmPartitionSettingsAsset, LandscapeData landscapeData, QualitySettingsAsset qualitySettingsAsset)
         {
             this.view = view;
 
-            // Clean current options loaded from the settings menu configuration and load names from QualitySettings
-            view.DropdownView.Dropdown.options.Clear();
-            foreach (string option in QualitySettings.names)
-                view.DropdownView.Dropdown.options.Add(new TMP_Dropdown.OptionData { text = option });
+            this.realmPartitionSettingsAsset = realmPartitionSettingsAsset;
+            this.landscapeData = landscapeData;
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs:26]: Same question here regarding landscapeData scope

File: Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs [hunk 3]
```diff
@@ -10,33 +14,96 @@ public class GraphicsQualitySettingsController : SettingsFeatureController
...
-            QualitySettings.SetQualityLevel(index);
+            if (index < view.DropdownView.Dropdown.options.Count - 1)
+                ForceSetQualityLevel(index);
+
             settingsDataStore.SetDropdownValue(GRAPHICS_QUALITY_DATA_STORE_KEY, index, save: true);
         }
 
+        private void ForceSetQualityLevel(int index)
+        {
+            int currentQualityLevel = QualitySettings.GetQualityLevel();
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs:60]: I am a bit confused at this logic, if we retrieve the current quality level and its the same why do we need to do any logic?

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                // Exclude possible duplicates
+                if (possibleResolutions.Contains(resolution))
+                    continue;
+
+                possibleResolutions.Add(resolution);
+                view.DropdownView.Dropdown.options.Add(new TMP_Dropdown.OptionData { text = $"{resolution.width}x{resolution.height} ({GetAspectRatio(resolution.width, resolution.height)}) {resolution.refreshRate} Hz" });
+            }
+        }
+
+        private static string GetAspectRatio(int width, int height)
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs:67]: This isn't returning the aspect ratio, its returning a string representing it, its probably cleaner to introduce a function that formats the dropdown with the right text. Notice earlier up we also calculate aspect ratio.

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            view.DropdownView.Dropdown.options.Clear();
+
+            // By design, we want the list to be inverted so the biggest resolutions stay on top
+            for (int index = Screen.resolutions.Length - 1; index >= 0; index--)
+            {
+                Resolution resolution = Screen.resolutions[index];
+
+                // Exclude all resolutions that are not 16:9 or 16:10
+                float aspectRatio = (float)resolution.width / resolution.height;
+                if (aspectRatio - (16.0f / 9.0f) != 0f && aspectRatio - (16.0f / 10.0f) != 0f)
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs:51]: Can this suffer from floating point comparison errors?

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        // By design, the default resolution should be a 1080p resolution with a 16:9 aspect ratio
+        private static bool IsDefaultResolution(Resolution resolution)
+        {
+            if (resolution.height != 1080)
+                return false;
+
+            float aspectRatio = (float)resolution.width / resolution.height;
+            return Mathf.Abs(aspectRatio - (16.0f / 9.0f)) == 0f;
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs:89]: Lets reduce the same calculations in this file and magic numbers

File: Explorer/Assets/DCL/Settings/ModuleControllers/SceneDistanceSettingsController.cs [hunk 7]
```diff
@@ -20,17 +20,23 @@ public SceneDistanceSettingsController(SettingsSliderModuleView view, RealmParti
...
+        private void OnMaxLoadingDistanceInParcelsChangedChanged(int newDistance)
+        {
+            view.SliderView.Slider.value = newDistance;
+            settingsDataStore.SetSliderValue(SCENE_DISTANCE_DATA_STORE_KEY, newDistance, save: true);
         }
 
         public override void Dispose()
         {
             view.SliderView.Slider.onValueChanged.RemoveListener(SetSceneDistanceSettings);
+            realmPartitionSettingsAsset.OnMaxLoadingDistanceInParcelsChanged -= OnMaxLoadingDistanceInParcelsChangedChanged;
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/SceneDistanceSettingsController.cs:39]: Similar to the landscape, can the scope of this change if we change realm?

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
+﻿using DCL.Settings.ModuleViews;
```
Comment (m3taphysics) [inline Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs:1]: I think in general this controller would benefit from moving some of its functionality to common utility and/or extention methods

File: Explorer/Assets/DCL/Settings/ModuleControllers/EnvironmentDistanceSettingsController.cs (hunk 1 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Settings/ModuleControllers/EnvironmentDistanceSettingsController.cs:25]: Anyone who changes the `landscapeData` scriptable object (that is how we're currently modifying the detail distance in out sustem), will make the slider control be updated. No matter from where this parameter is changed.

I linked this through an event in order to update the slider (among other cases) when we change the Quality preset dropdown.

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs (hunk 4 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs]: Refactor done here: https://github.com/decentraland/unity-explorer/pull/637/commits/188e12b414c22e8268bd6aaaa8a8cd92a489fa77

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs (hunk 5 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs]: Refactor done here: https://github.com/decentraland/unity-explorer/pull/637/commits/188e12b414c22e8268bd6aaaa8a8cd92a489fa77

File: Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs (hunk 6 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Settings/ModuleControllers/ResolutionSettingsController.cs]: Refactor done here: https://github.com/decentraland/unity-explorer/pull/637/commits/188e12b414c22e8268bd6aaaa8a8cd92a489fa77

File: Explorer/Assets/DCL/Settings/ModuleControllers/SceneDistanceSettingsController.cs (hunk 7 above)
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Settings/ModuleControllers/SceneDistanceSettingsController.cs:39]: Same explanation than [here](https://github.com/decentraland/unity-explorer/pull/637#discussion_r1574750569), but in this case with the `realmPartitionSettingsAsset` scriptable object.

File: Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs (hunk 2 above)
Comment (m3taphysics) [reply Explorer/Assets/DCL/Settings/ModuleControllers/GraphicsQualitySettingsController.cs:26]: Ok i see thanks!

**Review by m3taphysics — APPROVED** [review]

**Review by anicalbano — COMMENTED** [review]
🟡 Looks like a certain type of sound cannot be muted with any of the options. Is it because of the audio source?

https://github.com/decentraland/unity-explorer/assets/139157886/7aad5e4c-d44b-4827-8dd9-3ffdf0ff8296

**Review by Kinerius — APPROVED** [review]
LGTM!

**Review by anicalbano — APPROVED** [review]
🟢 Second check approved by QA

**Pull request merged**